### PR TITLE
BHBC-992-3: Update logout logic, Add dialog context

### PR DESCRIPTION
**Pull request by NickPhura**
# Overview

https://quartech.atlassian.net/browse/BHBC-992

## This PR contains the following changes

Updated the logout flow
- navigating to `/logout` will automatically kick off the logout process
- having a logout page allows the "Are you sure you want to cancel?" dialogs on the create pages to work, because they intercept navigation attempts
- should address Ola's testing comment.

Added a `DialogContext` context component.
- Currently supports the `ErrorDialog`, `YesNoDialog` and a new `Snackbar` component.
- Updated most of the existing uses of ErrorDialog and YesNoDialog to use the DialogContext version.
  - Left 1? in the Map Edit Controls component, because when I tried to change it, I was getting some bad behavior from the "Delete a shape" warning dialog. 
  - Haven't updated this for any of the new survey components.

## This PR contains the following types of changes

- [x] New feature (change which adds functionality)
- [x] Enhancement (improvements to existing functionality)
- [ ] Bug fix (change which fixes an issue)
- [ ] Misc cleanup / Refactoring / Documentation
- [ ] Version change

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [x] I have performed a self-review of my own code

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [x] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [x] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

Please add any relevant UI screenshots, if applicable.

### CODE REVIEW

File: app/src/features/projects/create/CreateProjectPage.tsx [hunk 1]
```diff
@@ -661,21 +658,36 @@ const CreateProjectPage: React.FC = () => {
...
+          history.push(location.pathname);
+        },
+        open: true
+      });
       return false;
     }
 
-    // If the cancel dialog is already open and a location change action is triggered by `handleDialogYes`: allow it
+    // If the cancel dialog is already open and another location change action is triggered: allow it
+    // This should only happen
```
Comment (sdevalapurkar) [inline app/src/features/projects/create/CreateProjectPage.tsx]: is this sentence supposed to continue and say more perhaps? `this should only happen....`

File: app/src/features/surveys/CreateSurveyPage.tsx [hunk 2]
```diff
@@ -252,15 +260,30 @@ const CreateSurveyPage = () => {
...
+          history.push(location.pathname);
+        },
+        open: true
+      });
       return false;
     }
 
-    // If the cancel dialog is already open and a location change action is triggered by `handleDialogYes`: allow it
+    // If the cancel dialog is already open and another location change action is triggered: allow it
+    // This should only happen
```
Comment (sdevalapurkar) [inline app/src/features/surveys/CreateSurveyPage.tsx]: same here

File: app/src/features/surveys/CreateSurveyPage.tsx [hunk 3]
```diff
@@ -252,15 +260,30 @@ const CreateSurveyPage = () => {
...
+      dialogContext.setYesNoDialog({
+        ...defaultCancelDialogProps,
+        onYes: () => {
+          dialogContext.setYesNoDialog({ open: false });
+          history.push(location.pathname);
+        },
+        open: true
+      });
       return false;
     }
```
Comment (sdevalapurkar) [inline app/src/features/surveys/CreateSurveyPage.tsx:283]: can this be moved to a shared file or something? it is used in many places

**Review by sdevalapurkar — COMMENTED** [review]
looking great, few small comments :) lmk what you think

**Review by anissa-agahchen — APPROVED** [review]
Tested in Openshift, looks good :-)

**Review by sdevalapurkar — APPROVED** [review]
download

**Pull request merged**